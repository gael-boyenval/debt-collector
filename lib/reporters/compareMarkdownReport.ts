const createTable = (data) => `
|File|Prev|Current|Trend|
|--|--|--|--|
${data
  .map((file) => `|${file.file}|${file.rev}|${file.current}|${file.trend}|`)
  .join('\n')}
`

const createFileTable = (fileResult) => `
<br/>
<br/>
<b>${fileResult.fileShortPath}</b><br/>
<br/>

|Broken rule|score|
|--|--|
${fileResult.brokenRules
  .map((rule) => `|üö´ ${rule.ruleTitle}|${rule.ruleTotalSore}|`)
  .join('\n')}
`

const getFileScoreComparaison = (data) => {
  let result = ''
  if (data.noChangesFiles.length > 0) {
    result += `
### üí§ Files with same debt :

${createTable(data.noChangesFiles)}

`
  }

  if (data.lessDeptFiles.length > 0) {
    result += `
### ‚úÖ Files with less debt :

${createTable(data.lessDeptFiles)}

`
  }

  if (data.moreDeptFiles.length > 0) {
    result += `
### ‚ùå Files with more debt :

${createTable(data.moreDeptFiles)}

`
  }

  return result
}

const getConclusions = (data) => {
  if (data.totalScores.solde > 0) {
    return `### ‚ùå Debt score for modified files increased by ${data.totalScores.solde} [^1]`
  }
  if (data.totalScores.solde < 0) {
    return `### ‚úÖ Debt score for modified files decreased by ${data.totalScores.solde} [^1]`
  }

  return '### üí§ Debt score for modified files did not change [^1]'
}

const getMotivationSpeatch = (data) => {
  if (data.totalScores.solde > 0) {
    return `Maybe try something else üò≠`
  }
  if (data.totalScores.solde < 0) {
    return `You did great ! üéâ`
  }

  return 'Neither good or bad, I guess ü§∑üèΩ'
}

const compareMarkDownReport = (data) => `
## Debt collector report

${getConclusions(data)}
${getMotivationSpeatch(data)}

|Previous debt|Current debt|trend|
|--|--|--|
|${data.totalScores.rev.toString()}|${data.totalScores.cur.toString()}|${data.totalScores.solde.toString()}|

<details>
<summary>
  <h3>Modified files ‚Ä¢ see scores before and after</h3>
</summary>
<div>

${getFileScoreComparaison(data)}

<br/>
<br/>
</div>
</details>


<details>
<summary>
  <h3>Help us improve code quality! Here are some ideas for you</h3>
</summary>
<div>

${data.currentResults
  .filter((res) => res.totalScore !== 0)
  .map(createFileTable)
  .join('\n')}

<br/>
<br/>
</div>
</details>

[^1]: Scores based on modified files only <br/>The report may not be accurate if your branch is not up to date with the base branch.
`

export default compareMarkDownReport
