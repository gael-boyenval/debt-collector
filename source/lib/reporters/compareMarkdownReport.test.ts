import { describe, it, expect } from 'vitest'
import compareMarkdownReport from './compareMarkdownReport.js'
import type { CompareResults } from '../types.js'

describe('compareMarkdownReport', () => {
  it('should handle case where all files have zero debt', () => {
    const data: CompareResults = {
      config: { include: [] },
      previousRevResult: {
        hash: 'abc',
        date: '2024-03-18',
        name: 'main',
        totalScore: 0,
        results: [],
      },
      currentRevResult: {
        hash: 'def',
        date: '2024-03-19',
        name: 'feature',
        totalScore: 0,
        results: [],
      },
      noChangesFiles: [],
      lessDeptFiles: [],
      moreDeptFiles: [],
    }

    expect(compareMarkdownReport(data)).toMatchSnapshot()
  })

  it('should handle increased debt score with multiple files', () => {
    const data: CompareResults = {
      config: { include: [] },
      previousRevResult: {
        hash: 'abc',
        date: '2024-03-18',
        name: 'main',
        totalScore: 20,
        results: [],
      },
      currentRevResult: {
        hash: 'def',
        date: '2024-03-19',
        name: 'feature',
        totalScore: 35,
        results: [
          {
            filePath: 'src/components/Button.tsx',
            fileShortPath: 'src/components/Button.tsx',
            totalScore: 10,
            brokenRules: [
              {
                ruleTitle: 'Complex Function',
                ruleId: 'complex-func',
                occurences: 3,
                ruleTotalSore: 15,
                debtScore: 5,
              },
            ],
          },
        ],
      },
      noChangesFiles: [],
      lessDeptFiles: [],
      moreDeptFiles: [
        {
          file: 'src/components/Button.tsx',
          rev: 10,
          current: 15,
          trend: 5,
        },
        {
          file: 'src/utils/helpers.ts',
          rev: 10,
          current: 20,
          trend: 10,
        },
      ],
    }

    expect(compareMarkdownReport(data)).toMatchSnapshot()
  })

  it('should handle decreased debt score with multiple files', () => {
    const data: CompareResults = {
      config: { include: [] },
      previousRevResult: {
        hash: 'abc',
        date: '2024-03-18',
        name: 'main',
        totalScore: 35,
        results: [
          {
            filePath: 'src/components/Button.tsx',
            fileShortPath: 'src/components/Button.tsx',
            totalScore: 10,
            brokenRules: [
              {
                ruleTitle: 'Complex Function',
                ruleId: 'complex-func',
                occurences: 3,
                ruleTotalSore: 15,
                debtScore: 5,
              },
            ],
          },
        ],
      },
      currentRevResult: {
        hash: 'def',
        date: '2024-03-19',
        name: 'feature',
        totalScore: 20,
        results: [],
      },
      noChangesFiles: [],
      lessDeptFiles: [
        {
          file: 'src/components/Button.tsx',
          rev: 15,
          current: 10,
          trend: -5,
        },
        {
          file: 'src/utils/helpers.ts',
          rev: 20,
          current: 10,
          trend: -10,
        },
      ],
      moreDeptFiles: [],
    }

    expect(compareMarkdownReport(data)).toMatchSnapshot()
  })

  it('should handle mixed changes with multiple broken rules', () => {
    const data: CompareResults = {
      config: { include: [] },
      previousRevResult: {
        hash: 'abc',
        date: '2024-03-18',
        name: 'main',
        totalScore: 30,
        results: [
          {
            filePath: 'src/components/Button.tsx',
            fileShortPath: 'src/components/Button.tsx',
            totalScore: 10,
            brokenRules: [
              {
                ruleTitle: 'Complex Function',
                ruleId: 'complex-func',
                occurences: 3,
                ruleTotalSore: 15,
                debtScore: 5,
              },
            ],
          },
        ],
      },
      currentRevResult: {
        hash: 'def',
        date: '2024-03-19',
        name: 'feature',
        totalScore: 35,
        results: [
          {
            filePath: 'src/components/Button.tsx',
            fileShortPath: 'src/components/Button.tsx',
            totalScore: 10,
            brokenRules: [
              {
                ruleTitle: 'Complex Function',
                ruleId: 'complex-func',
                occurences: 3,
                ruleTotalSore: 15,
                debtScore: 5,
              },
            ],
          },
        ],
      },
      noChangesFiles: [
        {
          file: 'src/types/index.ts',
          rev: 5,
          current: 5,
          trend: 0,
        },
      ],
      lessDeptFiles: [
        {
          file: 'src/components/Button.tsx',
          rev: 15,
          current: 10,
          trend: -5,
        },
      ],
      moreDeptFiles: [
        {
          file: 'src/utils/helpers.ts',
          rev: 10,
          current: 20,
          trend: 10,
        },
      ],
    }

    expect(compareMarkdownReport(data)).toMatchSnapshot()
  })

  it('should handle edge cases with empty arrays and zero scores', () => {
    const data: CompareResults = {
      config: { include: [] },
      previousRevResult: {
        hash: 'abc',
        date: '2024-03-18',
        name: 'main',
        totalScore: 0,
        results: [],
      },
      currentRevResult: {
        hash: 'def',
        date: '2024-03-19',
        name: 'feature',
        totalScore: 0,
        results: [],
      },
      noChangesFiles: [
        {
          file: 'src/components/Empty.tsx',
          rev: 0,
          current: 0,
          trend: 0,
        },
      ],
      lessDeptFiles: [],
      moreDeptFiles: [],
    }

    expect(compareMarkdownReport(data)).toMatchSnapshot()
  })

  it('should handle different file types and paths', () => {
    const data: CompareResults = {
      config: { include: [] },
      previousRevResult: {
        hash: 'abc',
        date: '2024-03-18',
        name: 'main',
        totalScore: 25,
        results: [],
      },
      currentRevResult: {
        hash: 'def',
        date: '2024-03-19',
        name: 'feature',
        totalScore: 30,
        results: [
          {
            filePath: 'src/components/Button.tsx',
            fileShortPath: 'src/components/Button.tsx',
            totalScore: 10,
            brokenRules: [
              {
                ruleTitle: 'Style Issues',
                ruleId: 'style-issues',
                occurences: 2,
                ruleTotalSore: 10,
                debtScore: 5,
              },
            ],
          },
        ],
      },
      noChangesFiles: [],
      lessDeptFiles: [
        {
          file: 'src/styles/main.css',
          rev: 10,
          current: 5,
          trend: -5,
        },
      ],
      moreDeptFiles: [
        {
          file: 'src/components/Header/Header.tsx',
          rev: 10,
          current: 15,
          trend: 5,
        },
        {
          file: 'src/pages/Home/index.tsx',
          rev: 5,
          current: 10,
          trend: 5,
        },
      ],
    }

    expect(compareMarkdownReport(data)).toMatchSnapshot()
  })
})
