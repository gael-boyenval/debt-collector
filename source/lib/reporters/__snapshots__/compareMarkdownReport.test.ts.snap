// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compareMarkdownReport > should handle case where all files have zero debt 1`] = `
"
## Debt collector report

All changed files have a debt score of 0.

Nothing to do here, we're all good ! 🎉
"
`;

exports[`compareMarkdownReport > should handle decreased debt score with multiple files 1`] = `
"
## Debt collector report

### ✅ Debt score for modified files decreased by -15 [^1]
You did great ! 🎉

|Previous debt|Current debt|trend|
|--|--|--|
|35|20|-15|

<details>
<summary>
  <h3>Modified files • see scores before and after</h3>
</summary>
<div>


### ✅ Files with less debt :


|File|Prev|Current|Trend|
|--|--|--|--|
|src/components/Button.tsx|15|10|-5|
|src/utils/helpers.ts|20|10|-10|




<br/>
<br/>
</div>
</details>


<details>
<summary>
  <h3>Help us improve code quality! Here are some ideas for you</h3>
</summary>
<div>



<br/>
<br/>
</div>
</details>

[^1]: Scores based on modified files only <br/>The report may not be accurate if your branch is not up to date with the base branch.
"
`;

exports[`compareMarkdownReport > should handle different file types and paths 1`] = `
"
## Debt collector report

### ❌ Debt score for modified files increased by 5 [^1]
Maybe try something else 😭

|Previous debt|Current debt|trend|
|--|--|--|
|25|30|5|

<details>
<summary>
  <h3>Modified files • see scores before and after</h3>
</summary>
<div>


### ✅ Files with less debt :


|File|Prev|Current|Trend|
|--|--|--|--|
|src/styles/main.css|10|5|-5|



### ❌ Files with more debt :


|File|Prev|Current|Trend|
|--|--|--|--|
|src/components/Header/Header.tsx|10|15|5|
|src/pages/Home/index.tsx|5|10|5|




<br/>
<br/>
</div>
</details>


<details>
<summary>
  <h3>Help us improve code quality! Here are some ideas for you</h3>
</summary>
<div>


<br/>
<br/>
<b>src/components/Button.tsx</b><br/>
<br/>

|Broken rule|score|
|--|--|
|🚫 Style Issues|10|


<br/>
<br/>
</div>
</details>

[^1]: Scores based on modified files only <br/>The report may not be accurate if your branch is not up to date with the base branch.
"
`;

exports[`compareMarkdownReport > should handle edge cases with empty arrays and zero scores 1`] = `
"
## Debt collector report

All changed files have a debt score of 0.

Nothing to do here, we're all good ! 🎉
"
`;

exports[`compareMarkdownReport > should handle increased debt score with multiple files 1`] = `
"
## Debt collector report

### ❌ Debt score for modified files increased by 15 [^1]
Maybe try something else 😭

|Previous debt|Current debt|trend|
|--|--|--|
|20|35|15|

<details>
<summary>
  <h3>Modified files • see scores before and after</h3>
</summary>
<div>


### ❌ Files with more debt :


|File|Prev|Current|Trend|
|--|--|--|--|
|src/components/Button.tsx|10|15|5|
|src/utils/helpers.ts|10|20|10|




<br/>
<br/>
</div>
</details>


<details>
<summary>
  <h3>Help us improve code quality! Here are some ideas for you</h3>
</summary>
<div>


<br/>
<br/>
<b>src/components/Button.tsx</b><br/>
<br/>

|Broken rule|score|
|--|--|
|🚫 Complex Function|15|


<br/>
<br/>
</div>
</details>

[^1]: Scores based on modified files only <br/>The report may not be accurate if your branch is not up to date with the base branch.
"
`;

exports[`compareMarkdownReport > should handle mixed changes with multiple broken rules 1`] = `
"
## Debt collector report

### ❌ Debt score for modified files increased by 5 [^1]
Maybe try something else 😭

|Previous debt|Current debt|trend|
|--|--|--|
|30|35|5|

<details>
<summary>
  <h3>Modified files • see scores before and after</h3>
</summary>
<div>


### 💤 Files with same debt :


|File|Prev|Current|Trend|
|--|--|--|--|
|src/types/index.ts|5|5|0|



### ✅ Files with less debt :


|File|Prev|Current|Trend|
|--|--|--|--|
|src/components/Button.tsx|15|10|-5|



### ❌ Files with more debt :


|File|Prev|Current|Trend|
|--|--|--|--|
|src/utils/helpers.ts|10|20|10|




<br/>
<br/>
</div>
</details>


<details>
<summary>
  <h3>Help us improve code quality! Here are some ideas for you</h3>
</summary>
<div>


<br/>
<br/>
<b>src/components/Button.tsx</b><br/>
<br/>

|Broken rule|score|
|--|--|
|🚫 Complex Function|15|


<br/>
<br/>
</div>
</details>

[^1]: Scores based on modified files only <br/>The report may not be accurate if your branch is not up to date with the base branch.
"
`;
